src/Npgsql.GeoJSON/GeoJSONHandler.cs

```diff
@@ -0,0 +1,798 @@
+﻿#region License
+// The PostgreSQL License
+//
+// Copyright (C) 2018 The Npgsql Development Team
+//
+// Permission to use, copy, modify, and distribute this software and its
+// documentation for any purpose, without fee, and without a written
+// agreement is hereby granted, provided that the above copyright notice
+// and this paragraph and the following two paragraphs appear in all copies.
+//
+// IN NO EVENT SHALL THE NPGSQL DEVELOPMENT TEAM BE LIABLE TO ANY PARTY
+// FOR DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES,
+// INCLUDING LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE AND ITS
+// DOCUMENTATION, EVEN IF THE NPGSQL DEVELOPMENT TEAM HAS BEEN ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//
+// THE NPGSQL DEVELOPMENT TEAM SPECIFICALLY DISCLAIMS ANY WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
+// AND FITNESS FOR A PARTICULAR PURPOSE. THE SOFTWARE PROVIDED HEREUNDER IS
+// ON AN "AS IS" BASIS, AND THE NPGSQL DEVELOPMENT TEAM HAS NO OBLIGATIONS
+// TO PROVIDE MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
+#endregion
+
+using GeoJSON.Net;
+using GeoJSON.Net.CoordinateReferenceSystem;
+using GeoJSON.Net.Geometry;
+using Npgsql.BackendMessages;
+using Npgsql.TypeHandling;
+using System;
+using System.Collections.ObjectModel;
+using System.Threading.Tasks;
+
+#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member
+
+namespace Npgsql.GeoJSON
+{
+    [Flags]
+    public enum GeoJSONOptions
+    {
+        None = 0,
+        BoundingBox = 1,
+        ShortCRS = 2,
+        LongCRS = 4
+    }
+
+    public sealed class GeoJSONHandlerFactory : NpgsqlTypeHandlerFactory<GeoJSONObject>
+    {
+        readonly GeoJSONOptions _options;
+
+        public GeoJSONHandlerFactory(GeoJSONOptions options = GeoJSONOptions.None)
+            => _options = options;
+
+        protected override NpgsqlTypeHandler<GeoJSONObject> Create(NpgsqlConnection conn)
+        {
+            var builder = new CrsMapBuilder();
+            if ((_options & (GeoJSONOptions.ShortCRS | GeoJSONOptions.LongCRS)) != GeoJSONOptions.None)
+                using (var cmd = new NpgsqlCommand(
+                    "SELECT min(srid), max(srid), auth_name " +
+                    "FROM(SELECT srid, auth_name, srid - rank() OVER(ORDER BY srid) AS range " +
+                    "FROM spatial_ref_sys) AS s GROUP BY range, auth_name ORDER BY 1;", conn))
+                using (var reader = cmd.ExecuteReader())
```

Comment:
The problem is here. If I add mappings both for `geometry` and `geography`, then `AddMapping` will be called twice. This means that the `Create` method will be called twice too, and the query above will be executed twice.

There are two ways to fix the problem. The first one is hacky and works only for pooled connections, but it allows to register both mappings:

```csharp
public static INpgsqlTypeMapper UseGeoJson(this INpgsqlTypeMapper mapper, GeoJSONOption option = GeoJSONOption.None, bool geographyAsDefault = false);
```

```csharp
CrsMap GetCrsMap(NpgsqlConnection conn)
{
    if ((_options & (GeoJSONOptions.ShortCRS | GeoJSONOptions.LongCRS)) == GeoJSONOptions.None)
        return default;

    if (s_crsMaps.TryGetValue(conn.ConnectionString, out var crsMap))
        return crsMap;

    var connString = new NpgsqlConnectionStringBuilder(conn.ConnectionString);
    if (connString.Pooling)
    {
        var builder = new CrsMapBuilder();
        using (var cmd = new NpgsqlCommand(
                "SELECT min(srid), max(srid), auth_name " +
                "FROM(SELECT srid, auth_name, srid - rank() OVER(ORDER BY srid) AS range " +
                "FROM spatial_ref_sys) AS s GROUP BY range, auth_name ORDER BY 1;", conn))
        using (var reader = cmd.ExecuteReader())
            while (reader.Read())
            {
                builder.Add(new CrsMapEntry(
                    reader.GetInt32(0),
                    reader.GetInt32(1),
                    reader.GetString(2)));
            }

        crsMap = builder.Build();

        lock (s_crsMaps)
            s_crsMaps.Add(conn.ConnectionString, crsMap);
    }

    return crsMap;
}
```

The second way isn't tricky, but it allows only one type mapping. It's possible to support two mappings at the same time, but then the query will be executed twice. The advantage here is that an user is responsible for his own actions.

```csharp
public static INpgsqlTypeMapper UseGeoJsonForGeography(this INpgsqlTypeMapper mapper, GeoJSONOption option = GeoJSONOption.None);
public static INpgsqlTypeMapper UseGeoJsonForGeometry(this INpgsqlTypeMapper mapper, GeoJSONOption option = GeoJSONOption.None);
```

Reply:
Done.

> One thing I don't understand is why pooling has to be turned on - why not cache the authorities regardless of pooling?

I don't like static caches when their lifetime depends on an object. It would be better to cache handlers with their data in the connection or in the related pool. So, it was a way to limit the authority map cache.

> And second, instead of taking a lock, why not just have a ConcurrentDictionary and use GetOrAdd() with a lambda that executes the query?

It was a sketch (:

Reply:
Oh ok :)

The code as it is now looks really OK to me - a single static ConcurrentDictionary cache should really be OK. I'm OK for merging.